llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp

```diff
@@ -948,6 +952,61 @@ bool RISCVInstructionSelector::selectFPCompare(MachineInstr &MI,
   return true;
 }
 
+bool RISCVInstructionSelector::selectIsFPClass(MachineInstr &MI,
+                                               MachineIRBuilder &MIB,
+                                               MachineRegisterInfo &MRI) const {
+  Register GISFPCLASS = MI.getOperand(0).getReg();
+  Register Src = MI.getOperand(1).getReg();
+  const MachineOperand &ImmOp = MI.getOperand(2);
+  const unsigned XLen = STI.getXLen();
+  unsigned NewOpc = MRI.getType(Src).getSizeInBits() == 32 ? RISCV::FCLASS_S
+                                                           : RISCV::FCLASS_D;
+
+  // Turn LLVM IR's floating point classes to that in RISC-V,
+  // by simply rotating the 10-bit immediate right by two bits.
+  APInt GFpClassImm(10, static_cast<uint64_t>(ImmOp.getImm()));
+  APInt FClassMask = GFpClassImm.rotr(2).zext(XLen);
+
+  Register FClassResult = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+  // Insert FCLASS_S/D.
+  auto FClass = MIB.buildInstr(NewOpc, {FClassResult}, {Src});
+  if (!FClass.constrainAllUses(TII, TRI, RBI))
+    return false;
+
+  if (FClassMask == 1) {
+    // We don't need to generate additional instructions if Mask is 1, as the
+    // `ANDI rs, 1` that follows will suffice.
```

Comment:
I think we have to make an ANDI. It's not guaranteed to exist. If computeKnownBits ran on the G_ISFPCLASS, it would know that G_ISFPCLASS produces 0/1 and delete any existing ANDI.

Reply:
I don't exactly follow why this needs post processing, but my feeling is if you're introducing regular bit instructions after this result you'd be better off doing this in custom legalization

Reply:
> I think we have to make an ANDI. It's not guaranteed to exist. If computeKnownBits ran on the G_ISFPCLASS, it would know that G_ISFPCLASS produces 0/1 and delete any existing ANDI.

Done

Reply:
> I don't exactly follow why this needs post processing, but my feeling is if you're introducing regular bit instructions after this result you'd be better off doing this in custom legalization

I think some contexts have been shifted around, which lines / post processing are you referring here?

Reply:
I think we should have a RISCV::G_FCLASS node and use custom legalization to turn G_FPCLASS into that plus G_AND and G_ICMP during legalization. Then we can use the existing isel patterns for RISCVISD::FPCLASS(which I think should be named RISCVISD::FCLASS so we should fix that too) by marking G_FCLASS as equivalent.